assembly/src/parsers/ast/stack_ops.rs

```diff
@@ -1,174 +1,219 @@
+///
+/// # Errors
+/// Returns an error if the instruction token contains wrong number of parameters, or if the
+/// provided parameter is greater than 3.
 pub fn parse_dupw(op: &Token) -> Result<Node, AssemblyError> {
-    let node = match op.num_parts() {
-        0 => return Err(AssemblyError::missing_param(op)),
-        1 => Node::Instruction(Instruction::DupW0),
+    debug_assert_eq!(op.parts()[0], "dupw");
+    match op.num_parts() {
+        0 => unreachable!(),
+        1 => Ok(Instruction(DupW0)),
         2 => match op.parts()[1] {
-            "0" => Node::Instruction(Instruction::DupW0),
-            "1" => Node::Instruction(Instruction::DupW1),
-            "2" => Node::Instruction(Instruction::DupW2),
-            "3" => Node::Instruction(Instruction::DupW3),
-            _ => return Err(AssemblyError::invalid_param(op, 1)),
+            "0" => Ok(Instruction(DupW0)),
+            "1" => Ok(Instruction(DupW1)),
+            "2" => Ok(Instruction(DupW2)),
+            "3" => Ok(Instruction(DupW3)),
+            _ => Err(AssemblyError::invalid_param(op, 1)),
         },
-        _ => return Err(AssemblyError::extra_param(op)),
-    };
-
-    Ok(node)
+        _ => Err(AssemblyError::extra_param(op)),
+    }
 }
 
+/// Returns `Swap` node instruction if no immediate vaule is provided or one of the
+/// `Swap` — `Swap15` instructions according to the immediate value.
+///
+/// # Errors
+/// Returns an error if the instruction token contains wrong number of parameters, or if the
+/// provided parameter is smaller than 1 or greater than 15.
 pub fn parse_swap(op: &Token) -> Result<Node, AssemblyError> {
-    validate_operation!(op, "swap", 0..1);
-
-    let node = match op.num_parts() {
-        1 => Node::Instruction(Instruction::Swap),
+    debug_assert_eq!(op.parts()[0], "swap");
+    match op.num_parts() {
+        0 => unreachable!(),
+        1 => Ok(Instruction(Swap)),
         2 => match op.parts()[1] {
-            "1" => Node::Instruction(Instruction::Swap),
-            "2" => Node::Instruction(Instruction::Swap2),
-            "3" => Node::Instruction(Instruction::Swap3),
-            "4" => Node::Instruction(Instruction::Swap4),
-            "5" => Node::Instruction(Instruction::Swap5),
-            "6" => Node::Instruction(Instruction::Swap6),
-            "7" => Node::Instruction(Instruction::Swap7),
-            "8" => Node::Instruction(Instruction::Swap8),
-            "9" => Node::Instruction(Instruction::Swap9),
-            "10" => Node::Instruction(Instruction::Swap10),
-            "11" => Node::Instruction(Instruction::Swap11),
-            "12" => Node::Instruction(Instruction::Swap12),
-            "13" => Node::Instruction(Instruction::Swap13),
-            "14" => Node::Instruction(Instruction::Swap14),
-            "15" => Node::Instruction(Instruction::Swap15),
-            _ => return Err(AssemblyError::invalid_param(op, 1)),
+            "1" => Ok(Instruction(Swap)),
```

Comment:
Since we are refactoring, should we follow the same pattern of `Dup` and rename it to `Swap1`?

Reply:
Good point! Will update in a follow-up PR.